api/label/set.go

```diff
@@ -0,0 +1,372 @@
+	if !id.Valid() {
+		// Invalid IDs are not cached.
+		return encoder.Encode(l.Iter())
+	}
+
+	var lookup *string
+	l.lock.Lock()
+	for idx := 0; idx < maxConcurrentEncoders; idx++ {
+		if l.encoders[idx] == id {
+			lookup = &l.encoded[idx]
+			break
+		}
+	}
+	l.lock.Unlock()
+
+	if lookup != nil {
+		return *lookup
+	}
+
+	r := encoder.Encode(l.Iter())
+
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	for idx := 0; idx < maxConcurrentEncoders; idx++ {
+		if l.encoders[idx] == id {
+			return l.encoded[idx]
+		}
+		if !l.encoders[idx].Valid() {
+			l.encoders[idx] = id
+			l.encoded[idx] = r
+			return r
+		}
+	}
+
+	// TODO: This is a performance cliff.  Find a way for this to
+	// generate a warning.
+	return r
+}
+
+// NewSet returns a new `*Set`.  Except for empty sets, this method
+// adds an additional allocation compared with a call to
+// `NewSetWithSortable`.  See that method for more details.
+func NewSet(kvs ...core.KeyValue) Set {
+	// Check for empty set.
+	if len(kvs) == 0 {
+		return emptySet
+	}
+
+	return NewSetWithSortable(kvs, new(Sortable))
+}
+
+// NewSetWithSortable modifies the input slice, ensuring the
+// following:
+//
+// - Last-value wins semantics
+// - Caller sees the reordering, but doesn't lose values
+// - Repeated call preserve last-value wins.
+//
+// Note that methods are defined `*Set`, although no allocation for
+// `Set` is required.  Callers may avoid memory allocations by:
+//
+// - allocating a `Sortable` for use as a temporary in this method
+// - allocating a `Set` for storing the return value of this
+//   constructor.
+func NewSetWithSortable(kvs []core.KeyValue, tmp *Sortable) Set {
+	// The requitements stated above require that the stable
+	// result be placed in the end of the input slice, while
+	// overwritten values are swapped to the beginning.
+
+	// Check for empty set.
+	if len(kvs) == 0 {
+		return emptySet
+	}
+
+	*tmp = kvs
+
+	// Sort and de-duplicate.  Note: this use of `tmp`
+	// avoids an allocation because it is a pointer.
+	sort.Stable(tmp)
+
+	*tmp = nil
+
+	position := len(kvs) - 1
+	offset := position - 1
+
+	for ; offset >= 0; offset-- {
+		if kvs[offset].Key == kvs[position].Key {
+			continue
+		}
+		kvs[offset], kvs[position-1] = kvs[position-1], kvs[offset]
+		position = position - 1
+	}
+
+	return Set{
+		equivalent: computeDistinct(kvs[position:]),
+	}
+}
+
+// computeDistinct returns a `Distinct` using either the fixed- or
+// reflect-oriented code path, depending on the size of the input.
+func computeDistinct(kvs []core.KeyValue) Distinct {
+	iface := computeDistinctFixed(kvs)
+	if iface == nil {
+		iface = computeDistinctReflect(kvs)
+	}
+	return Distinct{
+		iface: iface,
+	}
+}
+
+// computeDistinctFixed computes a `Distinct` for small slices.  It
+// returns nil if the input is too large for this code path.
+func computeDistinctFixed(kvs []core.KeyValue) interface{} {
+	switch len(kvs) {
+	case 1:
+		ptr := new([1]core.KeyValue)
+		copy((*ptr)[:], kvs)
```

Comment:
just a nit, but this duplicate code:

```
copy((*ptr)[:], kvs)
return *ptr
```

could be consolidated to follow the `switch` statement.

Reply:
Some of this could be consolidated, but ptr itself is of different type in each case.